File: src/slowql/rules/security/dos.py
"""
Security Dos rules.
"""

from __future__ import annotations

from typing import Any

from sqlglot import exp

from slowql.core.models import Category, Dimension, Fix, Issue, Query, Severity
from slowql.rules.base import ASTRule, PatternRule

__all__ = [
    "PgSleepUsageRule",
    "PgSleepUsageRule",
    "RegexDenialOfServiceRule",
    "TsqlWaitforDelayRule",
    "UnboundedRecursiveCTERule",
]


class UnboundedRecursiveCTERule(ASTRule):
    """Detects recursive CTEs without MAXRECURSION limits."""

    id = "SEC-DOS-001"
    name = "Unbounded Recursive CTE"
    description = (
        "Detects recursive CTEs without MAXRECURSION limits, which can consume unlimited resources."
    )
    severity = Severity.HIGH
    dimension = Dimension.SECURITY
    category = Category.SEC_DOS

    def check_ast(self, query: Query, ast: Any) -> list[Issue]:
        issues = []

        for node in ast.walk():
            # Check for WITH RECURSIVE or recursive CTE pattern
            if isinstance(node, exp.With):
                for cte in node.expressions:
                    if isinstance(cte, exp.CTE):
                        # Check if CTE references itself (recursive)
                        cte_name = getattr(cte, "alias", "")
                        cte_query = getattr(cte, "this", None)

                        if cte_name and cte_query and self._is_recursive(cte_query, cte_name):
                            # Check if OPTION (MAXRECURSION) exists in outer query
                            # This is a simplified check
                            query_str = query.raw.upper()
                            if "MAXRECURSION" not in query_str:
                                issues.append(
                                    self.create_issue(
                                        query=query,
                                        message=f"Recursive CTE '{cte_name}' without MAXRECURSION limit",
                                        snippet=str(cte)[:100],
                                        impact=(
                                            "Unbounded recursion can consume all available memory and CPU. "
                                            "A malicious recursive CTE can crash the database server or trigger cloud cost explosion."
                                        ),
                                        fix=Fix(
                                            description="Always set MAXRECURSION: OPTION (MAXRECURSION 100). Design recursion with guaranteed termination conditions.",
                                            replacement="",
                                            is_safe=False,
                                        ),
                                    )
                                )

        return issues

    def _is_recursive(self, query_node: Any, cte_name: str) -> bool:
        """Check if CTE references itself"""
        for node in query_node.walk():
            if isinstance(node, exp.Table):
                name = getattr(node, "name", "")
                if name.lower() == cte_name.lower():
                    return True
        return False


class RegexDenialOfServiceRule(PatternRule):
    """Detects regular expressions with patterns known to cause catastrophic backtracking."""

    id = "SEC-DOS-002"
    name = "Regex Denial of Service (ReDoS)"
    description = (
        "Detects regular expressions with patterns known to cause catastrophic backtracking."
    )
    severity = Severity.MEDIUM
    dimension = Dimension.SECURITY
    category = Category.SEC_DOS

    pattern = r"\b(REGEXP|RLIKE|REGEXP_LIKE|REGEXP_MATCHES|SIMILAR\s+TO)\s*\(?[^)]*(\(\?\:?\[?\w+\]\*\)[\*\+]|\(\.\*\)[\*\+]|\(\w\+\)[\*\+]|\[\^?\w+\]\*\[\^?\w+\]\*)"
    message_template = "Potential ReDoS pattern detected: {match}"

    impact = (
        "ReDoS patterns like (a+)+ or (.*)* can take exponential time on crafted input. "
        "A single malicious input can hang database threads for hours."
    )
    fix_guidance = "Use RE2-compatible patterns only (no backreferences, atomic groups). Set regex timeouts. Validate regex patterns before accepting user input."


class PgSleepUsageRule(PatternRule):
    """Detects pg_sleep() calls which may indicate DoS vectors or testing artifacts."""

    id = "SEC-PG-001"
    name = "pg_sleep Usage Detected"
    description = (
        "Detects calls to pg_sleep() which is a PostgreSQL-specific function. "
        "In production code this is almost always a testing artifact or a "
        "deliberate denial-of-service vector used to verify blind SQL injection."
    )
    severity = Severity.MEDIUM
    dimension = Dimension.SECURITY
    category = Category.SEC_DOS
    dialects = ("postgresql",)

    pattern = r"\bpg_sleep\s*\("
    message_template = "pg_sleep() call detected — potential DoS vector or testing artifact: {match}"

    impact = (
        "pg_sleep() ties up a database connection for the specified duration. "
        "An attacker can use it to confirm blind SQL injection or exhaust the "
        "connection pool, causing denial of service."
    )
    fix_guidance = (
        "Remove pg_sleep() calls from production code. If used for testing, "
        "guard behind a feature flag or test-only configuration."
    )


class TsqlWaitforDelayRule(PatternRule):
    """Detects WAITFOR DELAY in T-SQL production code."""

    id = "PERF-TSQL-004"
    name = "WAITFOR DELAY in Production Code"
    description = (
        "WAITFOR DELAY pauses execution for the specified duration. In "
        "production code this is almost always a testing artifact or a "
        "blind SQL injection indicator."
    )
    severity = Severity.MEDIUM
    dimension = Dimension.PERFORMANCE
    category = Category.PERF_EXECUTION
    dialects = ("tsql",)

    pattern = r"\bWAITFOR\s+DELAY\b"
    message_template = "WAITFOR DELAY detected — testing artifact or blind injection vector: {match}"

    impact = (
        "WAITFOR DELAY ties up a connection and worker thread for the "
        "specified duration. An attacker can use it to confirm blind "
        "SQL injection or exhaust the connection pool."
    )
    fix_guidance = (
        "Remove WAITFOR DELAY from production code. If used for polling, "
        "use Service Broker or Query Notification instead."
    )

File: slowql-rs/src/rules/security/dos.rs
use crate::models::issue::Category;
use crate::models::{Dimension, Issue, Query, Severity};
use crate::rules::base::{DialectSet, Rule};
use once_cell::sync::Lazy;
use regex::Regex;

struct UnboundedRecursiveCteRule;
static PAT_DOS_001: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(?i)\bWITH\s+RECURSIVE\b|\bWITH\b[\s\S]*?\bUNION\s+ALL\b").unwrap()
});

impl Rule for UnboundedRecursiveCteRule {
    fn id(&self) -> &'static str { "SEC-DOS-001" }
    fn name(&self) -> &'static str { "Unbounded Recursive CTE" }
    fn severity(&self) -> Severity { Severity::High }
    fn dimension(&self) -> Dimension { Dimension::Security }
    fn category(&self) -> Option<Category> { Some(Category::SecDos) }
    fn impact(&self) -> &'static str { "Unbounded recursion can consume all available memory and CPU, crashing the database server." }

    fn check(&self, query: &Query) -> Vec<Issue> {
        if !PAT_DOS_001.is_match(&query.raw) { return Vec::new(); }
        let raw_upper = query.raw_upper();
        if raw_upper.contains("MAXRECURSION") { return Vec::new(); }
        vec![self.build_issue(
            query,
            "Recursive CTE without MAXRECURSION limit - unbounded recursion risk",
            &query.raw[..query.raw.len().min(100)],
        )]
    }
}

struct RegexDenialOfServiceRule;
static PAT_DOS_002: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(?i)\b(REGEXP|RLIKE|REGEXP_LIKE|REGEXP_MATCHES|SIMILAR\s+TO)\s*\(?[^)]*(\(\?\:?\[?\w+\]\*\)[\*\+]|\(\.\*\)[\*\+]|\(\w\+\)[\*\+]|\[\^?\w+\]\*\[\^?\w+\]\*)").unwrap()
});

impl Rule for RegexDenialOfServiceRule {
    fn id(&self) -> &'static str { "SEC-DOS-002" }
    fn name(&self) -> &'static str { "Regex Denial of Service (ReDoS)" }
    fn severity(&self) -> Severity { Severity::Medium }
    fn dimension(&self) -> Dimension { Dimension::Security }
    fn category(&self) -> Option<Category> { Some(Category::SecDos) }
    fn impact(&self) -> &'static str { "ReDoS patterns like (a+)+ can take exponential time on crafted input, hanging database threads for hours." }

    fn check(&self, query: &Query) -> Vec<Issue> {
        PAT_DOS_002.find(&query.raw).map(|m| {
            vec![self.build_issue(query, &format!("Potential ReDoS pattern detected: {}", m.as_str()), m.as_str())]
        }).unwrap_or_default()
    }
}

struct PgSleepUsageRule;
static PAT_PG_001: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(?i)\bpg_sleep\s*\(").unwrap()
});

impl Rule for PgSleepUsageRule {
    fn id(&self) -> &'static str { "SEC-PG-001" }
    fn name(&self) -> &'static str { "pg_sleep Usage Detected" }
    fn severity(&self) -> Severity { Severity::Medium }
    fn dimension(&self) -> Dimension { Dimension::Security }
    fn category(&self) -> Option<Category> { Some(Category::SecDos) }
    fn dialects(&self) -> DialectSet { DialectSet::new(&["postgresql"]) }
    fn impact(&self) -> &'static str { "pg_sleep() ties up a database connection and can exhaust the connection pool, causing denial of service." }

    fn check(&self, query: &Query) -> Vec<Issue> {
        if !self.dialect_matches(query) { return Vec::new(); }
        PAT_PG_001.find(&query.raw).map(|m| {
            vec![self.build_issue(query, &format!("pg_sleep() call detected - potential DoS vector: {}", m.as_str()), m.as_str())]
        }).unwrap_or_default()
    }
}

struct TsqlWaitforDelayRule;
static PAT_TSQL_WAIT: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(?i)\bWAITFOR\s+DELAY\b").unwrap()
});

impl Rule for TsqlWaitforDelayRule {
    fn id(&self) -> &'static str { "PERF-TSQL-004" }
    fn name(&self) -> &'static str { "WAITFOR DELAY in Production Code" }
    fn severity(&self) -> Severity { Severity::Medium }
    fn dimension(&self) -> Dimension { Dimension::Performance }
    fn category(&self) -> Option<Category> { Some(Category::PerfExecution) }
    fn dialects(&self) -> DialectSet { DialectSet::new(&["tsql"]) }
    fn impact(&self) -> &'static str { "WAITFOR DELAY ties up a connection and worker thread, exhausting the connection pool under load." }

    fn check(&self, query: &Query) -> Vec<Issue> {
        if !self.dialect_matches(query) { return Vec::new(); }
        PAT_TSQL_WAIT.find(&query.raw).map(|m| {
            vec![self.build_issue(query, &format!("WAITFOR DELAY detected - testing artifact or blind injection vector: {}", m.as_str()), m.as_str())]
        }).unwrap_or_default()
    }
}

pub fn rules() -> Vec<Box<dyn Rule>> {
    vec![
        Box::new(UnboundedRecursiveCteRule),
        Box::new(RegexDenialOfServiceRule),
        Box::new(PgSleepUsageRule),
        Box::new(TsqlWaitforDelayRule),
    ]
}
